File: backend.scala/app/com/skewwhiffy/auraltester/controller/ScaleController.scala
package com.skewwhiffy.auraltester.controller

import com.skewwhiffy.auraltester.internalnotation.InternalNotationFactory
import com.skewwhiffy.auraltester.model.ScaleResponse
import com.skewwhiffy.auraltester.scales.{Key, ScaleDirection, ScaleTypeFactory}
import com.skewwhiffy.auraltester.services.{AbcService, ScaleService}
import play.api.libs.json.{Json, OFormat}
import play.api.mvc._

import javax.inject._
import scala.util.chaining.scalaUtilChainingOps

@Singleton
class ScaleController @Inject()(
  val controllerComponents: ControllerComponents,
  private val abcService: AbcService,
  private val internalNotationFactory: InternalNotationFactory,
  private val scaleService: ScaleService,
  private val scaleTypeFactory: ScaleTypeFactory
) extends BaseController {
  private implicit val scaleJson: OFormat[ScaleResponse] = Json.format[ScaleResponse]
  def index(
    clef: String,
    note: String,
    scaleType: String,
    direction: String
  ): Action[AnyContent] = Action {
    val clefObject = internalNotationFactory.clef(clef)
    val noteObject = internalNotationFactory.getNote(note).note
    val (scaleTypeObject, isMinor) = scaleType match {
      case "major" => (scaleTypeFactory.major, false)
      case "minor-harmonic" => (scaleTypeFactory.minorHarmonic, true)
      case "minor-melodic" => (scaleTypeFactory.minorMelodic, true)
      case _ => throw new IllegalArgumentException(s"Unrecognized scale type: '$scaleType'")
    }
    val directionObject = direction match {
      case "ascending" => ScaleDirection.ascending
      case "descending" => ScaleDirection.descending
      case _ => throw new IllegalArgumentException(s"Unrecognized direction: '$direction'")
    }
    val scale = scaleService.getScale(clefObject, noteObject, scaleTypeObject, directionObject)
    val key = scale.lowestNote.note.pipe(it => Key(it, isMinor))
    val response = ScaleResponse(abcService.getAbc(clefObject, scale), abcService.getAbc(clefObject, scale, key))
    Ok(Json.toJson(response))
  }
}

File: backend.old/src/main/kotlin/com/skewwhiffy/auraltester/controller/ScaleController.kt
package com.skewwhiffy.auraltester.controller

import com.skewwhiffy.auraltester.dto.ScaleResponse
import com.skewwhiffy.auraltester.notation.factory.InternalNotationFactory
import com.skewwhiffy.auraltester.notation.factory.ScaleTypeFactory
import com.skewwhiffy.auraltester.notation.model.key.Key
import com.skewwhiffy.auraltester.notation.model.scale.ScaleDirection
import com.skewwhiffy.auraltester.service.AbcService
import com.skewwhiffy.auraltester.service.ScaleService
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RestController

@RestController
class ScaleController(
    private val abcService: AbcService,
    private val internalNotationFactory: InternalNotationFactory,
    private val scaleService: ScaleService,
    private val scaleTypeFactory: ScaleTypeFactory
) {

    @GetMapping("/api/scale")
    fun get(clef: String, note: String, scaleType: String, direction: String): ScaleResponse {
        val clefObject = internalNotationFactory.clef(clef)
        val noteObject = internalNotationFactory.getNote(note).note
        val scaleTypeAndMinor = when (scaleType) {
            "major" -> scaleTypeFactory.major to false
            "minor-harmonic" -> scaleTypeFactory.minorHarmonic to true
            "minor-melodic" -> scaleTypeFactory.minorMelodic to true
            else -> throw IllegalArgumentException("Unrecognized scale type: $scaleType")
        }
        val scaleTypeObject = scaleTypeAndMinor.first
        val isMinor = scaleTypeAndMinor.second
        val directionObject = when (direction) {
            "ascending" -> ScaleDirection.ASCENDING
            "descending" -> ScaleDirection.DESCENDING
            else -> throw IllegalArgumentException("Unrecognized direction: '$direction'")
        }
        val scale = scaleService.getScale(clefObject, noteObject, scaleTypeObject, directionObject)
        val lowestNote = scale.lowestNote.note
        val key = if (isMinor) Key.minor(lowestNote) else Key.major(lowestNote)
        return ScaleResponse(
            abcService.getAbc(clefObject, scale).abc,
            abcService.getAbc(clefObject, scale, key).abc
        )
    }
}
